lib/Dialect/D2M/IR/D2MOps.cpp

```diff
@@ -1206,6 +1213,41 @@ ::mlir::LogicalResult d2m::GenericOp::verify() {
                  (rankedTensorType && rankedTensorType.getEncoding());
   SmallVector<AffineMap> indexingMaps = getIndexingMapsValue();
   if (hasGrid && !indexingMaps.empty()) {
+    // Validate that all operands have device layouts before calling
+    // getOperandGridShapes(), which assumes layouts are present
+    for (Value operand : getOperands()) {
+      auto result =
+          llvm::TypeSwitch<Type, LogicalResult>(operand.getType())
+              .Case<MemRefType>([&](MemRefType memrefType) -> LogicalResult {
+                if (!mlir::dyn_cast<ttcore::DeviceLayoutInterface>(
+                        memrefType.getLayout())) {
+                  return emitOpError("memref operand must have a device layout "
+                                     "attribute "
+                                     "(e.g., #ttcore.shard, #ttcore.view, or "
+                                     "#ttcore.interleaved), "
+                                     "but got: ")
+                         << memrefType;
+                }
+                return success();
+              })
+              .Case<RankedTensorType>(
+                  [&](RankedTensorType tensorType) -> LogicalResult {
+                    if (!mlir::dyn_cast_or_null<ttcore::MetalLayoutAttr>(
+                            tensorType.getEncoding())) {
+                      return emitOpError("tensor operand must have a metal "
+                                         "layout encoding, "
+                                         "but got: ")
+                             << tensorType;
+                    }
+                    return success();
+                  })
+              .Default([](Type) { return success(); });
+
+      if (failed(result)) {
+        return failure();
+      }
+    }
+
```

Comment:
Not related to the main goal of this PR, but did encounter verifier crashes while debugging things, which prompted this addition.